File: C++/227.basic-calculator-ii.cpp
/*
 * @lc app=leetcode id=227 lang=cpp
 *
 * [227] Basic Calculator II
 *
 * https://leetcode.com/problems/basic-calculator-ii/description/
 *
 * algorithms
 * Medium (37.02%)
 * Likes:    1556
 * Dislikes: 274
 * Total Accepted:    197K
 * Total Submissions: 530.8K
 * Testcase Example:  '"3+2*2"'
 *
 * Implement a basic calculator to evaluate a simple expression string.
 * 
 * The expression string contains only non-negative integers, +, -, *, /
 * operators and empty spaces  . The integer division should truncate toward
 * zero.
 * 
 * Example 1:
 * 
 * 
 * Input: "3+2*2"
 * Output: 7
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: " 3/2 "
 * Output: 1
 * 
 * Example 3:
 * 
 * 
 * Input: " 3+5 / 2 "
 * Output: 5
 * 
 * 
 * Note:
 * 
 * 
 * You may assume that the given expression is always valid.
 * Do not use the eval built-in library function.
 * 
 * 
 */
#include<vector>
#include<string>
using namespace std;

// @lc code=start
class Solution {
public:
    int calculate(string s) {
        long num = 0, n = s.size();
        char op = '+';
        
        vector<int> stack;
        for (int i=0; i<n; ++i) {
            char ch = s[i];
            if (isdigit(ch)) {
                num = num * 10 + s[i] - '0';
            }
            
            if (ch=='+' || ch=='-' || ch=='*' || ch=='/' || i == n - 1) {
                if (op == '+') stack.push_back(num);
                if (op == '-') stack.push_back(-num);
                if (op == '*' || op == '/') {
                    int tmp = (op=='*') ? stack.back() * num : stack.back() / num;
                    stack.pop_back();
                    stack.push_back(tmp);
                }
                op = ch;
                num = 0;
            } 
        }
        int ans = 0;
        while (!stack.empty()) {
            ans += stack.back();
            stack.pop_back();
        }
        return ans;
    }
};
// @lc code=end

File: Python/227.basic-calculator-ii.py
#
# @lc app=leetcode id=227 lang=python3
#
# [227] Basic Calculator II
#
# https://leetcode.com/problems/basic-calculator-ii/description/
#
# algorithms
# Medium (43.71%)
# Likes:    6140
# Dislikes: 842
# Total Accepted:    654.1K
# Total Submissions: 1.5M
# Testcase Example:  '"3+2*2"'
#
# Given a string s which represents an expression, evaluate this expression and
# return its value. 
# 
# The integer division should truncate toward zero.
# 
# You may assume that the given expression is always valid. All intermediate
# results will be in the range of [-2^31, 2^31 - 1].
# 
# Note: You are not allowed to use any built-in function which evaluates
# strings as mathematical expressions, such as eval().
# 
# 
# Example 1:
# Input: s = "3+2*2"
# Output: 7
# Example 2:
# Input: s = " 3/2 "
# Output: 1
# Example 3:
# Input: s = " 3+5 / 2 "
# Output: 5
# 
# 
# Constraints:
# 
# 
# 1 <= s.length <= 3 * 10^5
# s consists of integers and operators ('+', '-', '*', '/') separated by some
# number of spaces.
# s represents a valid expression.
# All the integers in the expression are non-negative integers in the range [0,
# 2^31 - 1].
# The answer is guaranteed to fit in a 32-bit integer.
# 
# 
#

# @lc code=start
class Solution:
    def calculate(self, s: str) -> int:
        num = 0
        op = '+'

        stack = []
        for i, c in enumerate(s):
            if c.isdigit():
                num = num * 10 + int(c)
            
            if c in "+-*/" or i == len(s)-1:
                if op == '+':
                    stack.append(num)
                elif op == '-':
                    stack.append(-num)
                elif op == '*':
                    stack[-1] = stack[-1] * num
                elif op == '/':
                    stack[-1] = int(stack[-1] / num)
                op = c
                num = 0
        return sum(stack)
# @lc code=end

